### DISCUSSION TITLE
Checking for block inside container blocks

### DISCUSSION THREAD

**Discussion by stephencottontail** [WordPress/gutenberg#40172] [Q&A]
I made a table of contents block by getting all the blocks with `getBlocks()` and then picking out all the `core/heading` blocks. Now I’ve got a situation where a client is adding heading blocks inside `core/group` or `core/column` blocks and since `getBlocks()` doesn’t check `innerBlocks`, those heading blocks don’t go into the ToC. Been wracking my brain trying to think of how to get at these blocks and was hoping that someone here would have an idea

edit: Forgot to mention that there could be heading blocks inside of a container block inside of another containing block so a solution capable of going arbitrarily deep would be nice

**Comment by talldan** [comment]
There's a table of contents block that's being worked on in Gutenberg (I think it's pretty close to being shipped).

It supports nested headings, so you might find inspiration in the code. This is the pull request with the latest changes to the block - https://github.com/WordPress/gutenberg/pull/29739.

It looks like it uses `getClientIdsWithDescendants`, which returns the client ids for all blocks in the post (possibly in a flat array), and iterates through those using `getBlockName( clientId )` to determine whether the block is a header.
